include/envoy/network/listener.h

```diff
@@ -114,6 +114,46 @@ class ListenerCallbacks {
   virtual void onNewConnection(ConnectionPtr&& new_connection) PURE;
 };
 
+/**
+ * UDP listener callbacks.
+ */
+class UdpListenerCallbacks {
+public:
+  enum class ErrorCode { SYSCALL_ERROR, UNKNOWN_ERROR };
+
+  virtual ~UdpListenerCallbacks() = default;
+
+  /**
+   * On first packet received on a UDP socket. This allows the callback handler
+   * to establish filter chain (or any other prepararion).
+   *
+   * @param local_address Local bound socket network address.
+   * @param peer_address Network address of the peer.
+   * @param data Data buffer received.
+   */
+  virtual void onNewConnection(Address::InstanceConstSharedPtr local_address,
+                               Address::InstanceConstSharedPtr peer_address,
+                               Buffer::InstancePtr&& data) PURE;
+  /**
+   * Called whenever data is received by the underlying Udp socket.
+   *
+   * @param local_address Local bound socket network address.
+   * @param peer_address Network address of the peer.
+   * @param data Data buffer received.
+   */
+  virtual void onData(Address::InstanceConstSharedPtr local_address,
```

Comment:
Relaying a comment originally made by ianswett@chromium.org on an earlier listener sketch, which also applies here: from the perspective of QUIC (which will eventually use this), it would be desirable to bundle (local_address, peer_address, data) in a Datagram or Packet object, to accomodate possible future addition of other attributes, e.g. TTL. E.g., something like [this](https://gist.github.com/mpwarres/188e71dce411577d63505bde49b4a71c).

Reply:
+1

Reply:
@mpwarres , @ianswett  Do you think we should have it here or in the QUIC filter layer? Do other UDP filters need the `Datagram`/`Packet` concept?

Reply:
Seems fine to just make a struct now to carry the data if we want that? That will avoid a lot of code site changes later to add new fields.

Reply:
+1, let's totally make it a struct sooner rather than later.